sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -170,21 +170,35 @@ class Analyzer(catalog: Catalog,
    * Replaces [[UnresolvedRelation]]s with concrete relations from the catalog.
    */
   object ResolveRelations extends Rule[LogicalPlan] {
-    def getTable(u: UnresolvedRelation) = {
+    def getTable(u: UnresolvedRelation, cteRelations: Map[String, LogicalPlan]) = {
       try {
-        catalog.lookupRelation(u.tableIdentifier, u.alias)
+          // In hive, if there is same table name in database and CTE definition,
+          // hive will use the table in database, not the CTE one.
+          // Taking into account the reasonableness and the implementation complexity,
+          // here use the CTE definition first, check table name only and ignore database name
+          cteRelations.get(u.tableIdentifier.last)
+            .map(relation => u.alias.map(Subquery(_, relation)).getOrElse(relation))
+            .getOrElse(catalog.lookupRelation(u.tableIdentifier, u.alias))
       } catch {
         case _: NoSuchTableException =>
           u.failAnalysis(s"no such table ${u.tableIdentifier}")
       }
     }
 
-    def apply(plan: LogicalPlan): LogicalPlan = plan transform {
-      case i @ InsertIntoTable(u: UnresolvedRelation, _, _, _) =>
+    def apply(plan: LogicalPlan): LogicalPlan = {
+      val cteRelations = new scala.collection.mutable.HashMap[String, LogicalPlan]()
+
+      plan transform {
```

Comment:
Can the with ever be anywhere other than the top?  I think it would be better to avoid using `transform` for side effects.  

Instead, maybe:

```scala
val (realPlan, cteRelations) = plan match {
  case With(child, relations) => (child, relations)
  case other => (other, Map.empty)
}
```

Either that or walk the plan with a `collect` and then do the `transform`.